kernel/alarm.c

```diff
@@ -0,0 +1,251 @@
+}
+
+int k_alarm_next_deadline_(u32_t *deadline)
+{
+	int rv = -1;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	if (!sys_dlist_is_empty(&ready_list)) {
+		rv = 0;
+		goto unlock;
+	}
+
+	sys_dnode_t *hp = sys_dlist_peek_head(&scheduled_list);
+
+	if (hp) {
+		struct k_alarm *ap = CONTAINER_OF(hp, struct k_alarm, node);
+		*deadline = ap->deadline;
+		rv = 1;
+	}
+unlock:
+	k_spin_unlock(&lock, key);
+	return rv;
+}
+
+int k_alarm_split_(u32_t now)
+{
+	int rv = 0;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	sys_dnode_t *cp = sys_dlist_peek_head(&scheduled_list);
+
+	while (cp) {
+		struct k_alarm *cap = alarm_due_by(cp, now);
+
+		if (!cap) {
+			break;
+		}
+		cap->state = K_ALARM_READY;
+		++rv;
+		cp = sys_dlist_peek_next_no_check(&scheduled_list, cp);
+	}
+	if ((rv == 0)
+	    && !sys_dlist_is_empty(&ready_list)) {
+		rv = -1;
+	}
+
+	if (cp) {
+		sys_dlist_split(&ready_list, &scheduled_list, cp);
+	} else {
+		sys_dlist_join(&ready_list, &scheduled_list);
+	}
+
+	k_spin_unlock(&lock, key);
+	return rv;
+}
+
+void k_alarm_process_ready_(void)
+{
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	do {
+		sys_dnode_t *cp = sys_dlist_get(&ready_list);
+
+		if (!cp) {
+			break;
+		}
+
+		struct k_alarm *cap = CONTAINER_OF(cp, struct k_alarm, node);
+
+		cap->state = K_ALARM_ACTIVE;
+		if (cap->handler_fn) {
+			k_spin_unlock(&lock, key);
+			cap->handler_fn(cap, cap->user_data);
+			key = k_spin_lock(&lock);
+		}
+		if (cap->state == K_ALARM_ACTIVE) {
+			cap->state = K_ALARM_UNSCHEDULED;
+		}
+	} while (true);
+	k_spin_unlock(&lock, key);
+}
+
+int z_impl_k_alarm_schedule(struct k_alarm *alarm,
+			    u32_t deadline,
+			    u32_t flags)
+{
+	int rv = -EINVAL;
+	bool update_deadline = false;
+
+	if (!alarm) {
+		goto out;
+	}
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	bool may_resched = true
+		&& (alarm->state == K_ALARM_SCHEDULED)
+		&& (flags & (K_ALARM_FLAG_REPLACE | K_ALARM_FLAG_IF_SOONER));
+
+	if (sys_dnode_is_linked(&alarm->node)
+	    && !may_resched) {
+		rv = -EINVAL;
+		goto unlock;
+	}
+	if ((alarm->state != K_ALARM_UNSCHEDULED)
+	    && (alarm->state != K_ALARM_ACTIVE)
+	    && (alarm->state != K_ALARM_CANCELLED)
+	    && !may_resched) {
+		rv = -EBUSY;
+		goto unlock;
+	}
+
+	u32_t now = k_cycle_get_32();
+	s32_t delay = deadline - now;
+
+	if (may_resched) {
+		if (K_ALARM_FLAG_REPLACE & flags) {
+			/* Unconditional reschedule */
+		} else if ((K_ALARM_FLAG_IF_SOONER & flags)
+			   && (delay >= (s32_t)(alarm->deadline - now))) {
+			/* Not sooner; keep unchanged */
+			rv = 1;
+			goto unlock;
+		}
+		sys_dlist_remove(&alarm->node);
+	}
+
+	alarm->deadline = deadline;
+	if (delay > 0) {
+		alarm->state = K_ALARM_SCHEDULED;
+		update_deadline = link_alarm(&scheduled_list, alarm, now);
+		rv = 1;
+	} else if (K_ALARM_FLAG_ERROR_IF_LATE & flags) {
+		rv = -EINVAL;
+	} else {
+		alarm->state = K_ALARM_READY;
+		update_deadline = link_alarm(&ready_list, alarm, now);
```

Comment:
Huh?  No, the ready list isn't what's used to calculate the deadline.  Once it's on the ready list it will be processed at the next opportunity, without affecting the compare register.

Reply:
i referred the following scenario:
`z_impl_k_alarm_schedule` is called late. alarm is linked to `ready_list` and state is set to `READY`. `update_deadline` is set to true and `z_alarm_update_deadline` is called. in `alarm_update_deadline_` `rc=0` so `compare==now` but `RTC->EVENTS_COMPARE[0]` is not set. We end up with `compare = now + COUNTER_MIN_DELTA` being set in `CC` register. RTC interrupt will happen 4 ticks from now and alarm will be triggered from there.

Reply:
OK, I understand the case now.  Thanks for explaining it.

Yes, the alarm will fire late.  This is no different from a sequence where the alarm was scheduled in the future and has reached its deadline by the time the counter is set; or when the delay occurred because something blocked the timer ISR for several ticks.

These are edge cases, and I believe they're handled in a reasonable way with reasonable complexity.

I would love it if there were some way to force a compare event to occur immediately, but AFAICT the ARM/Nordic peripherals don't allow that, so the only solution is to set the compare event to the shortest value we can be sure will fire without lapping.  (MSP430 does allow triggering interrupts manually, which made things like this really easy to deal with when I was working in that world.)

Reply:
what if instead of calling `alarm_update_deadline_` in case of readiness interrupt would be triggered (`NVIC_SetPending`). Or that can be part of `alarm_update_deadline_`

Reply:
That could trigger the ISR, but is there a way to force `EVENTS_COMPARE[0]` to be set?

This approach would work if we only used the ISR to process `CC[0]`.  I would rather not limit it that way, as I can see use cases for low-latency dedicated callbacks triggered by system time that would use the other channels and bypass the alarm infrastructure, as a Nordic-specific extension (e.g. this *might* eliminate the need for the Bluetooth stack to maintain its own RTC).

But it is a potential solution if there's a strong consensus that the delay in this case is unacceptable.  I personally don't consider it a serious problem that's worth excluding low-latency time-triggered callbacks.